### DISCUSSION TITLE
vkt camera rotation/transformation

### DISCUSSION THREAD

**Discussion by jsaintauret** [marcomusy/vedo#788] [Q&A] [answered]
Hello,

I came across this video below, and I would like to know if any one of you knows how I could do this kind of predefined camera rotation/transformation ?
https://user-images.githubusercontent.com/52704901/212181574-d7facbf0-5437-4870-a994-1046cd50af6d.mp4

The main idea That I'm trying like to implement is a way to constraint the render camera axis of rotation when I move the mouse without having to click on "ctrl", similar to what maltab does when you move a 3D plot with the mouse

![image](https://user-images.githubusercontent.com/52704901/212182538-9d4b0419-cf7b-4899-9cba-9885f6b63693.png)

thank you in advance

**Comment by ZiguoAtGitHub** [comment]
I think you should add one call_back() function bind to mouse move event.
In the call_back() function you can keep axis-z vertical with setting rotx and roty to zero.

**Reply by jsaintauret** [reply]
Hi thanks four your answer, I think I get the idea but could you please be more specific with the function I need to use? are rotx and rotz callable vtkcamera function ?

**Comment by marcomusy** [comment]
yes, there are various ways of doing it... I would first check if any modes works for you out-of-the-box in `show(..., mode=...)`.

- using a callback as Ziguo suggested eg

```python
from vedo import *

def func(event):
    k = event.keypress
    if k=="Right":
        plt.camera.Azimuth(1)
    elif k=="Left":
        plt.camera.Azimuth(-1)
    elif k=="Down":
        plt.camera.Elevation(1)
    elif k=="Up":
        plt.camera.Elevation(-1)
    elif k=="r":
        plt.reset_viewup()
        return
    plt.renderer.ResetCameraClippingRange()
    plt.render()

settings.enable_default_keyboard_callbacks = False
settings.enable_default_mouse_callbacks = False

world = Box(size=(30,15,8)).wireframe()
airplane = Mesh(dataurl+"cessna.vtk").c("green")

plt = Plotter(axes=1)
plt.add_callback("keypress", func)

plt.show(world, airplane, mode="image")
```
(this uses keypress events)

Or
```python
from vedo import *

def func(event):
    d = event.delta2d
    if d[0] and d[1]:
        return
    if d[0]:
        plt.camera.Azimuth(-d[0])
    if d[1]:
        plt.camera.Elevation(-d[1])
    plt.renderer.ResetCameraClippingRange()
    plt.render()

# settings.enable_default_keyboard_callbacks = False
settings.enable_default_mouse_callbacks = False

world = Box(size=(30,15,8)).wireframe()
airplane = Mesh(dataurl+"cessna.vtk").c("green")

plt = Plotter(axes=1)
plt.add_callback("Interaction", func)

plt.show(world, airplane, mode="image")
```
(this uses mouse events)

- (more difficult) write a custom https://vtk.org/doc/nightly/html/classvtkInteractorStyle.html

**Reply by jsaintauret** [reply]
Hi thank you very much for all the tips , I'll try all that and let you know ;)

**Reply by ZiguoAtGitHub** [reply]
> yes, there are various ways of doing it... I would first check if any modes works for you out-of-the-box in `show(..., mode=...)`.
> 
> * using a callback as Ziguo suggested eg
> 
> ```python
> from vedo import *
> 
> def func(event):
>     k = event.keypress
>     if k=="Right":
>         plt.camera.Azimuth(1)
>     elif k=="Left":
>         plt.camera.Azimuth(-1)
>     elif k=="Down":
>         plt.camera.Elevation(1)
>     elif k=="Up":
>         plt.camera.Elevation(-1)
>     elif k=="r":
>         plt.reset_viewup()
>         return
>     plt.renderer.ResetCameraClippingRange()
>     plt.render()
> 
> settings.enable_default_keyboard_callbacks = False
> settings.enable_default_mouse_callbacks = False
> 
> world = Box(size=(30,15,8)).wireframe()
> airplane = Mesh(dataurl+"cessna.vtk").c("green")
> 
> plt = Plotter(axes=1)
> plt.add_callback("keypress", func)
> 
> plt.show(world, airplane, mode="image")
> ```
> 
> (this uses keypress events)
> 
> Or
> 
> ```python
> from vedo import *
> 
> def func(event):
>     d = event.delta2d
>     if d[0] and d[1]:
>         return
>     if d[0]:
>         plt.camera.Azimuth(-d[0])
>     if d[1]:
>         plt.camera.Elevation(-d[1])
>     plt.renderer.ResetCameraClippingRange()
>     plt.render()
> 
> # settings.enable_default_keyboard_callbacks = False
> settings.enable_default_mouse_callbacks = False
> 
> world = Box(size=(30,15,8)).wireframe()
> airplane = Mesh(dataurl+"cessna.vtk").c("green")
> 
> plt = Plotter(axes=1)
> plt.add_callback("Interaction", func)
> 
> plt.show(world, airplane, mode="image")
> ```
> 
> (this uses mouse events)
> 
> * (more difficult) write a custom https://vtk.org/doc/nightly/html/classvtkInteractorStyle.html

Awesome reply!

**Reply by jsaintauret** [reply]
There is a singularity happening when elevation get to 90° it's call "Gimbal Lock" and the plot will have weird behavior. To fix it quickly I have found on internet that you could use `plt.camera.OrthogonalizeViewUp() ` after the `plt.camera.Elevation(-d[1])` as a workaround.

**Comment by jsaintauret** [comment]
Could you please tell me what delta2d stands for ?

**Reply by marcomusy** [reply]
it's the shift between the previous and current mouse position in 2d units.

**Comment by RubendeBruin** [comment, marked as answer]
I have a pretty complete custom interactor here: https://github.com/RubendeBruin/DAVE/blob/master/src/DAVE/visual_helpers/vtkBlenderLikeInteractionStyle.py

which does a lot of things, among others keeping the Z-axis vertical.

I still have to contribute this back to vtk but havent found the time to do so. Feel free to use/abuse as you wish.

**Reply by jsaintauret** [reply]
Hi thank you very much for this, I'm going to try it but so far after playing with it looks quite promising.
I still have some question though, the behavior that I'm looking for is mainly coming from mouse  MiddleButton press event to rotate the actor while keeping the Z axis vertical, so what I did is to rename that event with the mouse leftbutton event , is this the best way to do it ?
Also you said that the Z axis stays vertical by default or do I need to activate it ?

Marco have also one question for you, I tried to set the interactorstyle with the BlenderStyle class from Dave file:

```
self.vtk_widget = QVTKRenderWindowInteractor(self)
self.vtk_widget.SetInteractorStyle(style)
self.vp = Plotter(qt_widget=self.vtk_widget, bg='white')
```

 but it looks like the render windows always keep the default one TrackballCamera after I call `plotter.show()` later on on my code, it looks like default mode overwrite what I set earlier. Any advice to set an interactor style?

thanks

**Reply by jsaintauret** [reply]
Or may be I just set the InteractorStyle after calling the show() function like this:

```
            self.vp.show(self.mesh,axes=self.axes,camera=cam)
            style = BlenderStyle()
            self.vp.interactor.SetInteractorStyle(style)
```

See the reslult, looks very nice

https://user-images.githubusercontent.com/52704901/214944044-0fef4254-d375-4ea7-9899-ae9b70dd135e.mp4

**Reply by marcomusy** [reply]
> Feel free to use/abuse as you wish.

I abused it by adding it to https://github.com/marcomusy/vedo/blob/master/vedo/interactor_modes.py 
;-)
seems to do the job from what i can see!
```python
from vedo import *
settings.enable_default_keyboard_callbacks = False
settings.enable_default_mouse_callbacks = False
mode = interactor_modes.BlenderStyle()
mesh = Mesh(dataurl+"cow.vtk")
plt = Plotter()
plt.show(mesh, axes=1, mode=mode)
```

**Reply by jsaintauret** [reply]
Oh very nice!! I'll take a look when I have some time are your going to push it on the next release ? The only thing that I would suggest maybe is that the user could select the preferred control button, in my case I tweak Dave's interact or style to have the rotation/elevation/azimuth control with the left click button on the mouse it's seams more intuitive as opposed to have it with the middle button which is usually the pan zoom.

Would it be hard to have configure control style option?

**Reply by marcomusy** [reply]
One could do that by passing an option keyword eg something like
`mode = interactor_modes.BlenderStyle(use_left_button=True)`
but I cannot implement it  - you will need to make a PR :)

**Reply by jsaintauret** [reply]
Ok I'll keep that in mind

**Reply by RubendeBruin** [reply]
Thanks for the gentle reminder :-) Got me motivated to start on the merge request.
Hopefully it will one day be merged into vtk: https://gitlab.kitware.com/vtk/vtk/-/merge_requests/10128
